### PR TITLE
fix: bootstrap merge order

### PR DESCRIPTION
**Pull request by Morriz**
fixes redkubes/unassigned-issues#362

### CODE REVIEW

**Comment by Morriz** [comment]
I tested also team creation, so `env.gotmpl` change is known to work @j-zimnowoda

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
```

2021-12-20T16:28:06.609Z otomi:hfValues:info No teams or cluster info found. ENV_DIR is potentially empty.
2021-12-20T16:28:06.610Z otomi:bootstrap:info Intalling artifacts from otomi/core:v0.14.55
2021-12-20T16:28:06.612Z otomi:bootstrap:info Copied bin files
2021-12-20T16:28:06.617Z otomi:bootstrap:info Copied vscode folder
2021-12-20T16:28:06.741Z otomi:bootstrap:info Stored loose YAML schema at: /home/app/stack/env/.vscode/values-schema.yaml
2021-12-20T16:28:06.743Z otomi:bootstrap:log Copying skeleton files
2021-12-20T16:28:06.788Z otomi:bootstrap:log Loading chart values from /secret/values.yaml
2021-12-20T16:28:06.788Z otomi:bootstrap:info Checking if secret/default/otomi-generated-passwords already exists
2021-12-20T16:28:06.962Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/settings.yaml
2021-12-20T16:28:06.963Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/secrets.settings.yaml, skipping...
2021-12-20T16:28:06.964Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/cluster.yaml
2021-12-20T16:28:06.964Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/cert-manager.yaml, skipping...
2021-12-20T16:28:06.965Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/external-dns.yaml, skipping...
2021-12-20T16:28:06.968Z otomi:values:writeValues:info All values were written to ENV_DIR
2021-12-20T16:28:06.968Z otomi:generateSecrets:info Extracting secrets
2021-12-20T16:28:06.971Z otomi:generateSecrets:info First round of templating
2021-12-20T16:28:06.983Z otomi:generateSecrets:info Parsing values for second round of templating
2021-12-20T16:28:06.984Z otomi:generateSecrets:info Gather all values for the second round of templating
2021-12-20T16:28:06.985Z otomi:generateSecrets:info Second round of templating
2021-12-20T16:28:07.079Z otomi:generateSecrets:info Generated all secrets
2021-12-20T16:28:07.081Z otomi:createCustomCA:info Generating custom root CA
2021-12-20T16:28:07.361Z otomi:createCustomCA:info Generated CA key pair
2021-12-20T16:28:07.374Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/settings.yaml, skipping...
2021-12-20T16:28:07.375Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/secrets.settings.yaml, skipping...
2021-12-20T16:28:07.375Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/cluster.yaml, skipping...
2021-12-20T16:28:07.376Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/cert-manager.yaml, skipping...
2021-12-20T16:28:07.376Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/external-dns.yaml, skipping...
2021-12-20T16:28:07.377Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.drone.yaml
2021-12-20T16:28:07.378Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.gitea.yaml
2021-12-20T16:28:07.379Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.harbor.yaml
2021-12-20T16:28:07.380Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.keycloak.yaml
2021-12-20T16:28:07.381Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.kubeapps.yaml
2021-12-20T16:28:07.381Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.loki.yaml
2021-12-20T16:28:07.382Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.oauth2-proxy.yaml
2021-12-20T16:28:07.384Z otomi:values:writeValues:info All values were written to ENV_DIR
2021-12-20T16:28:07.862Z otomi:global:error Error: in /home/app/stack/helmfile.tpl/helmfile-dump-all.yaml: error during ../helmfile.d/snippets/derived.gotmpl.part.0 parsing: template: stringTemplate:73:63: executing "stringTemplate" at <$v.otomi.adminPassword>: map has no entry for key "adminPassword"
    at hfCore (/home/app/stack/dist/src/common/hf.js:46:30)
    exit code: 1
    at ChildProcess.<anonymous> (/home/app/stack/node_modules/zx/dist/bundle.cjs:15343:22)
    at ChildProcess.emit (node:events:390:28)
    at maybeClose (node:internal/child_process:1064:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
```

**Comment by Morriz** [comment]
fixed with latest commit..test added

**Comment by Morriz** [comment]
actually, what I fixed was the fact that cli bootstrap was writing derived values...I could not reproduce your error...are you sure you are testing with the right image?

**Comment by Morriz** [comment]
1. installed with chart and activated drone
2. created team + service and deployed
3. pulled values and bootstrapped

no issues

File: src/cmd/bootstrap.test.ts [hunk 1]
```diff
@@ -202,6 +202,11 @@ describe('Bootstrapping values', () => {
       beforeEach(() => {
         deps.isChart = false
       })
+      it('should retrieve previous user input when cluster provider is set', async () => {
+        deps.loadYaml.mockReturnValue({ ...values, cluster: { provider: 'set' } })
+        await processValues(deps)
+        expect(deps.hfValues).toHaveBeenCalledWith({ filesOnly: true })
```
Comment (j-zimnowoda) [inline src/cmd/bootstrap.test.ts:208]: why don't you check values that processValues returns?

File: chart/otomi/Chart.yaml [hunk 2]
```diff
@@ -4,7 +4,7 @@ description: A Helm chart for installing otomi in Kubernetes
 home: https://otomi.io/
 icon: https://otomi.io/img/otomi-logo.svg
 type: application
-version: 0.3.CHART_PATCH_PLACEHOLDER
+version: 0.3.0 #CHART_PATCH_PLACEHOLDER
```
Comment (j-zimnowoda) [inline chart/otomi/Chart.yaml:7]: CHART_PATCH_PLACEHOLDER is not needed anymore?

File: chart/otomi/templates/post-job.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+              echo "Otomi has been uninstalled."
+          env:
+            - name: CI
+              value: '1'
+            - name: ENV_DIR
+              value: /home/app/stack/env
+            - name: VERBOSITY
+              value: '1'
+          resources:
+            limits:
```
Comment (j-zimnowoda) [inline chart/otomi/templates/post-job.yaml:39]: why such a high requests if it is just performing Kubectl command?

File: src/common/values.ts [hunk 4]
```diff
@@ -36,11 +36,11 @@ export const getK8sVersion = (): string => {
  * Find what image tag is defined in configuration for otomi
  * @returns string
  */
-export const getImageTag = async (): Promise<string> => {
+export const getImageTag = async (isBootstrap = false): Promise<string> => {
   if (process.env.OTOMI_TAG) return process.env.OTOMI_TAG
+  if (isBootstrap) return `v${pkg.version}`
```
Comment (j-zimnowoda) [inline src/common/values.ts:41]: what if the values indicate a certain version?

File: helmfile.d/snippets/defaults.yaml [hunk 5]
```diff
@@ -51,6 +51,9 @@ environments:
             enabled: true
             adminUsername: admin
             theme: otomi
+            idp:
```
Comment (j-zimnowoda) [inline helmfile.d/snippets/defaults.yaml:54]: could you clarify why these defaults are needed?

File: helmfile.d/snippets/derived.gotmpl [hunk 6]
```diff
@@ -87,4 +87,5 @@ environments:
             enabled: false #{{ eq $provider "aws" }}
         otomi:
           version: {{ $otomiTag }}
-        versions: {{- $versions | toYaml | nindent 10 }} 
+        versions: {{- $versions | toYaml | nindent 10 }}
+      - ../core.yaml
```
Comment (j-zimnowoda) [inline helmfile.d/snippets/derived.gotmpl:91]: could you clarify why core.yaml is now needed?

**Comment by Morriz** [comment]
Good question. Remnants of the previous destroy command I think. It needed it

File: src/cmd/bootstrap.test.ts (hunk 1 above)
Comment (Morriz) [reply src/cmd/bootstrap.test.ts:209]: it is not what is being tested, and we just stubbed the user input

Comment (Morriz) [reply src/cmd/bootstrap.test.ts:209]: also, the coverage of the parts in processValues after is made by the rest of the tests. Check the coverage.

File: chart/otomi/Chart.yaml (hunk 2 above)
Comment (Morriz) [reply chart/otomi/Chart.yaml:7]: oops...hate that mechanism...have to find a workaround

File: src/common/values.ts (hunk 4 above)
Comment (Morriz) [reply src/common/values.ts:41]: there are no values when bootstrap asks for it and this is only for reporting feedback, so trivial

Comment (Morriz) [reply src/common/values.ts:41]: no user shall see discrepancy between release and dev image, only developers...

File: helmfile.d/snippets/derived.gotmpl (hunk 6 above)
Comment (Morriz) [reply helmfile.d/snippets/derived.gotmpl:91]: I moved it to the next phase as it is not user input

File: helmfile.d/snippets/defaults.yaml (hunk 5 above)
Comment (Morriz) [reply helmfile.d/snippets/defaults.yaml:54]: they're just defaults and that section always exists as keycloak is mandatory

**Pull request merged**